google_guest_agent/network/manager/wicked_linux.go

```diff
@@ -0,0 +1,178 @@
+	configDir string
+}
+
+const (
+	// defaultWickedConfigDir is the default location for wicked configuration files.
+	defaultWickedConfigDir = "/etc/sysconfig/network"
+
+	// wickedCommand is the expected path to the wicked CLI.
+	wickedCommand = "/usr/sbin/wicked"
+)
+
+// init adds this network manager service to the list of known managers.
+func init() {
+	registerManager(&wicked{
+		configDir: defaultWickedConfigDir,
+	}, false)
+}
+
+// Name returns the name of this network manager service.
+func (n wicked) Name() string {
+	return "wicked"
+}
+
+// IsManaging checks whether wicked is managing the provided interface.
+func (n wicked) IsManaging(ctx context.Context, iface string) (bool, error) {
+	// Check the current main network service. Primarily applicable to SUSE images.
+	res := run.WithOutput(ctx, "systemctl", "status", "network.service")
+	if strings.Contains(res.StdOut, "wicked.service") {
+		return true, nil
+	}
+
+	// Check if the wicked service is running.
+	res = run.WithOutput(ctx, "systemctl", "is-active", "wicked.service")
+	if res.ExitCode != 0 {
+		return false, nil
+	}
+
+	// Check the status of configured interfaces.
+	res = run.WithOutput(ctx, wickedCommand, "ifstatus", iface, "--brief")
+	if res.ExitCode != 0 {
+		return false, fmt.Errorf("failed to check status of wicked configuration: %s", res.StdErr)
+	}
+	fields := strings.Fields(res.StdOut)
+	if fields[1] == "up" || fields[1] == "setup-in-progress" {
+		return true, nil
+	}
+	return false, nil
+}
+
+// Setup writes the necessary configuration files for each interface and enables them.
+func (n wicked) Setup(ctx context.Context, cfg *cfg.Sections, payload []metadata.NetworkInterfaces) error {
+	if len(payload) < 2 {
+		return nil
+	}
+	ifaces, err := interfaceNames(payload)
+	if err != nil {
+		return fmt.Errorf("failed to get network interfaces: %v", err)
+	}
+
+	if err = writeWickedConfigs(n.configDir, ifaces[1:]); err != nil {
+		return fmt.Errorf("error writing wicked configurations: %v", err)
+	}
+
+	args := append([]string{"ifup", "--timeout", "1"}, ifaces[1:]...)
+	return run.Quiet(ctx, wickedCommand, args...)
+}
+
+// writeWickedConfigs writes config files for the given ifaces in the given configuration
+// directory.
+func writeWickedConfigs(configDir string, ifaces []string) error {
+	var priority = 10100
+
+	// Write the config for all the non-primary network interfaces.
+	for _, iface := range ifaces {
+		logger.Debugf("write enabling ifcfg-%s config", iface)
+
+		var ifcfg *os.File
+
+		ifcfg, err := os.Create(ifcfgFilePath(configDir, iface))
+		if err != nil {
+			return err
+		}
+		defer ifcfg.Close()
+		contents := []string{
+			googleComment,
+			"STARTMODE=hotplug",
+			// NOTE: 'dhcp' is the dhcp4+dhcp6 option.
+			"BOOTPROTO=dhcp",
+			fmt.Sprintf("DHCLIENT_ROUTE_PRIORITY=%d", priority),
+		}
+		_, err = ifcfg.WriteString(strings.Join(contents, "\n"))
+		if err != nil {
+			return err
+		}
+		priority += 100
```

Comment:
why do we need incremented priority per interface again?

Reply:
I've been snooping around some documentation, but I haven't found anything concrete. But in [this example](https://www.suse.com/support/kb/doc/?id=000021237), they increment the priority by 100 past the first interface. This is mainly to maintain parity with the old implementation, but I agree that it seems pretty arbitrary at the moment.

Reply:
Based on my understanding we want only to have eth0 < eth1 < ethN. It seems to be common practice to increase by 100 but to me is looks more like a range reservation rather than anything else. It also connects with the initial value of 10100, it could be 100 or 1.